text/0069-api-v2.md

```diff
@@ -0,0 +1,195 @@
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes but only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+#### RawKV Value Encode
+
+If the key has RawKV prefix, which are `k{keyspace id prefix}r` or `k{keyspace id prefix}a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+#### TTL in Titan
+
+Titan is a storage engine plugin for RocksDB that store the value of a kvpair in a separated blob file if the value is large enough. This brings a new challenge for RawKV TTL compaction filter where the compaction filter can not know the ttl without reading the blob file. So with Titan, the RawKV value encode is defined to be:
+
+##### Value
+
+1. `{0x0}`
+2. `{0x1}{TTL expire timestamp}`
+
+##### Blob Value
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### Keyspace Management
+
+Add a new http interface to PD for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    description: "this is foo",
+    id: "0"
+  },
+  {
+    name: "redis",
+    id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "foo",
+  // optional
+  description: "this is foo"
+}
+
+// delete a keyspace
+DELETE /keyspaces/{keyspace_name}
+{}
+
+// recover the latest deleted keyspace
+POST /keyspaces?action=flashback
+{
+  new_name: "bar",
+}
+```
+
+The keyspaces are stored in etcd and has the no limitation on the id number.
+
+Note that when deleting a keyspace, the data in the keysapce will not be deleted automatically, the user should clean up the keyspace before deleting the kepspace. If a keyspace is deleted by accident, the user should be able to recover the keyspace by creating a keyspace using the same id as before. When a keyspace is deleted, the client should aware of the deletion in 5 mins, and after 5 mins, it guranetees that no client should be using the keyspace.
+
+### How to safely enable API V2
+
+#### Upgrade
+
+Upgrade from `API V1` to `API V2` is a simple process:
+
+1. Update TiKV, TiDB, and PD to the version that supports `API V2`.
+2. Ensure that no non-TiDB data is in TiKV. Delete if any.
+3. Ensure that no non-TiDB client is online. (manual process)
+4. Use `pd-ctl` to enable `API V2`.
+5. Enable `API V2` in TiKV config file and restart TiKV.
```

Comment:
this is a heavy step that breaks the existing session. could we enable API v2 without restarting the tikv-servers?

Reply:
We have to do it

Reply:
why?